WIP - [image-registry] Automate Pruning

adambkaplan opened the pull request:
Automate the image pruning maintenance task for the registry.
This is a necessary addition to ensure an OpenShift cluster remains healthy.

/assign @bparees @derekwaynecarr 

/cc @dmage @coreydaley @ricardomaraschini @smarterclayton

adambkaplan commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,252 @@
+    schedule: "*/1 * * * *"
+    keepTagRevisions: 3
+    keepYoungerThan: 60m
+    resources: {}
```
Something to consider - setting the resources for the pruner based on cluster info (number of `Image` objects, for example).

I don't think this is a requirement for GA, but would be a good feature to add to the backlog.

adambkaplan replied:
Added to JIRA: https://jira.coreos.com/browse/DEVEXP-441

bparees commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,252 @@
+1. A `ServiceAccount` dedicated to the image pruner.
+2. A `ClusterRoleBinding` granting the `system:image-pruner` role to the pruner
+service account.
+3. A `CronJob` which runs the image pruner with the following tunable parameters:
```
I wouldn't expect the operator to directly install the cronjob, i'd expect it to bootstrap a pruning configuration (possibly part of the registry operator resource) on initial install, and then i'd expect the operator to take that pruning configuration and produce/reconcile/update a cronjob to accomplish the requested pruning config.

dmage replied:
@bparees I don't understand the different between "directly install the cronjob" and "produce/reconcile/update a cronjob", could you elaborate on this?

bparees replied:
it might just be the way i was parsing the description, but the section says:

`The image registry operator will be updated to install the following:`

Some of the things it lists sound like things i'd expect to be in the manifest for the operator (thus created by the CVO when the operator is installed/created.  e.g. service accounts, rolebindings).

But the CronJob definition would not be in the manifests...  i would expect the cronjob to be created/managed by the operator in the same way the the operator creates/manages the registry deployment.

So i'd expect the flow to be:

1) CVO installs the manifest resources (service accounts, roles/bindings, operator deployment, etc)

2) operator comes up, creates its own registry config resource (e.g. the registry storage configuration today) which would now include configuration for automated pruning  (what we call bootstrapping its config)

3) operator sees its configuration, which would include config for automated pruning, and create a cronjob w/ the appropriate args (and on updates to that configuration, would sync any changes to args to that cronjob).

adambkaplan replied:
@bparees updated language to make it clear it's the _operator's_ job to create and reconcile, not the CVO:
```
The image registry operator will be updated to configure and reconcile all the
components needed to run image pruning on an OpenShift cluster.
```

bparees replied:
well, the SA and rolebindings will be reconciled by the CVO assuming you create them as manifest resources.

but i think the current phrasing covers the key part which is how the cronjob is managed.

bparees commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,252 @@
+### Upgrade / Downgrade Strategy
+
+On new installs and upgrades, the pruner will be installed on the cluster.
+For GA the pruner will be enabled and run daily at midnight.
```
it's only going to be enabled automatically on new installs, not upgrades, right?

dmage replied:
I don't like the idea that new clusters and upgraded clusters have different default configuration. For some customers it's important to have the pruner enabled.

@bparees would you prefer to have an alert instead of enabling pruning during upgrades?

bparees replied:
I don't like having different defaults either, but we definitely cannot just turn pruning on on someone's cluster as part of an upgrade, we might destroy a bunch of images.

and we definitely want to have this on in general, which means turning it on by default when we can (new installs).

All things considered i don't think having some clusters w/ pruning enabled and some with it not enabled is the worst thing in the world..  since admins can always turn it off anyway, we could/will still end up in that situation regardless of how we default new installs+upgrades.

Having an alert that pruning is not enabled on the cluster seems reasonable to me, to encourage those admins to enable it.

adambkaplan replied:
@bparees added separate section for upgrades.

adambkaplan commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,266 @@
+spec:
+  ...
+  pruning:
+    schedule: "*/1 * * * *"
```
@bparees  @openshift/openshift-team-developer-experience  I've struggled with this part of the API.

We initially discussed having the `schedule` field be what determines if the CronJob is enabled or suspended. However it may be more prudent to have a separate field `enabled` that is the primary flag to set. `schedule` could then default to "daily at midnight" if `enabled: true` and schedule is left empty.

dmage replied:
What about adding a ManagementState? It actually can be "Managed", "Unmanaged" and "Removed".

dmage replied:
@adambkaplan @bparees it's the other way around. I envision the pruner should have its own configuration resource and the operator should manage them independently. The pruner is not part of the registry, it's more about the image API and if someone decides to remove the registry from their cluster, they still may need to have pruner. For example, an image stream may become big because of scheduled imports. So I like the idea of having a separate operator for pruning (and we already have the repository cluster-image-pruner-operator).

gabemontero replied:
at first blush I agree with @dmage here .... mixing more than one config obj (this new one, and the image registry operator's existing one) under the auspices of a single operator seems messy

If the pruner needs to some how interface with the image registry operator and what it manages, those specifics should be called out / understood now and then exposed in the image registry operator's config object if need be

bparees replied:
While I can see the value in that distinction, please keep in mind the implementation cost of maintaining yet another operator, image, repo, CI jobs, ART process, etc.

Do we think a lot of users are going to use imagestreams but not the registry?  Can we live with telling that subset of users they are on their own to do pruning?  The cases where people blow up their cluster w/ images, as far as i know, are cases where they are using our registry and keep building+pushing new images to it.

Alternatively you can have a separate CRD for pruning if you want, without creating a separate operator.  Of course then you get into weirdness where the registry CRD says "removed" but they still have a valid pruning configuration that the operator is processing, and the telemeter status for the operator would report "removed" even though in reality it's still doing some stuff, so that feels a bit ugly.

You could of course also implement the actual "operator" logic in the openshift-controller (again, introducing a separate CRD for configuring automated pruning).  There's really no reason it needs its own explicit CVO-managed operator.  The main benefit of having a CVO-managed operator is easy telemeter reporting of whether it's healthy or not, imho.

adambkaplan replied:
> Do we think a lot of users are going to use imagestreams but not the registry? Can we live with telling that subset of users they are on their own to do pruning? 

@siamaksade any insights here?

adambkaplan replied:
@openshift/api-reviewers if we declare a field as Tech Preview on a `v1` API object, can we remove it in a subsequent release? Such as:

1. 4.3 add `pruning:` as a tech preview field on `config.imageregistry.operator.openshift.io`
2. 4.4 move pruning API to a new GA CRD (`config.imagepruner.operator.openshift.io`) - deprecate `pruning` object in image registry operator CRD.
3. 4.5 delete `pruning` from image registry operator CRD.

siamaksade replied:
@adambkaplan there are customers who use imagestreams without the internal registry but they are not the majority. I'm not aware of more granular data on this. That said, I don't think we should target pruning external registries yet. The asks we currently have are for pruning the internal registry as @bparees stated.

bparees replied:
the question isn't about pruning external registries.  it's about pruning image metadata we've imported from external registries, because it takes up space in your etcd, even if it's not taking up space in your registry storage.

siamaksade replied:
In case of the internal registry, auto-pruning removes also the images as I understand. Is that correct?

bparees replied:
first let's fix our terminology.

There are two critical pieces of data:

1) image/imagestream resource objects in etcd.  These exist both for images stored in the internal registry, as well as when you import an image from an external registry (which again, just means importing the metadata, not the actual image).

2) image content/blobs stored in the registry storage.  These only exist for images that were pushed to the internal registry, or that the internal registry did a "pullthrough" on.

When you run manual pruning (oc adm prune) you can choose whether you want to clean up (1), (2), or (1+2).

Automated pruning is not, i do not think, going to give you that choice.  It will always clean up (1+2).

In the case of external images, where we only have (1), then only (1) will be cleaned up, of course.

adambkaplan replied:
> Automated pruning is not, i do not think, going to give you that choice. It will always clean up (1+2).

This was something the team discussed at scrum today. If we separate the pruner operator from the registry operator, we can be more intelligent (though still take away the choice):

1. If registry is `Managed`, 1+2 for items on the internal registry, 1 for external images
2. If registry is `Removed`, 1 only (everything should be an external image).

Bundling the pruner as a component installed by the registry operator makes is harder to build an API that lets us have "1 only".

bparees replied:
you don't have to separate the operator to accomplish that.

i don't even think you have to separate the api, strictly speaking (no reason you can't set the management state to removed, but then have pruning configured, within the same CR..it's a bit weird, but you get to choose the meaning/implications).

Anyway "removed" refers to the operand.  In the case of this particular CR, the "operand" is the registry.

There is no reason we can't have a single operator with multiple operands (the registry being one, the pruner being the other).

Heck, you can even have multiple ClusterVersion objects being created/managed by your single operator. (which would allow you to report degraded status for pruning independent of degraded status for the registry).

So I suggest you go with a distinct CRD for pruning configuration.  Whether that CRD has the traditional "managementstate" field or not, i'm fairly indifferent but leaning towards not having it.

And whether you want to create an additional ClusterVersion object for reporting the health of the "pruning" operand, i'm also largely indifferent, i'd suggest deferring that for now.

I'd start by just firing alerts when pruning goes wrong.

adambkaplan replied:
Another reason for having a separate CRD for the pruner - it gives us the flexibility to land this feature as Tech Preview. Per @deads2k we can't have tech preview fields on GA APIs.

I think we've reached consensus that this API should live in a separate CRD - I'll update the enhancement.

adambkaplan commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,266 @@
+For GA the pruner will be enabled and run daily at midnight.
+
+On upgrade the cluster admin will be responsible for setting the schedule for
+the pruning `CronJob`. Doing so implicitly enables the pruner. If pruning is
```
...unless we add an `enabled` field to the API. Then IMO that is what enables the pruner.

dmage commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,266 @@
+
+1. `concurrencyPolicy: Forbid` - this prevents concurrent cron jobs from
+running.
+2. `suspend` - based on the provided `schedule`. If empty `suspend = true`.
```
We can just delete or skip creation of CronJob if the pruner is disabled.

adambkaplan replied:
I'm less a fan of that - doing so would remove any retained history from the child `Job`/`Pod` objects.

adambkaplan commented:
@bparees @dmage @openshift/openshift-team-developer-experience comments addressed and commits squashed. ptal

gabemontero commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,267 @@
+
+### Implementation Details/Notes/Constraints
+
+The implementation relies the current image pruning process invoked by the
```
`relies the current image pruning process` -> `relies on the current image pruning process`

i.e. add the `on1

gabemontero commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,267 @@
+    tolerations: {}
+    history:
+      successfulJobsHistoryLimit: 3
+      failedJobsHistoryLimit: 1
```
I'm late to the party, so apologies if this came up before.

Will the config object here maintain any status conditions .... perhaps around jobs in progress?  Or history?  Or does has the CronJob status been deemed sufficient for this?

adambkaplan replied:
@gabemontero good point on conditions. We haven't discussed these yet.

I'm open to suggestions, but IMO we need one to indicate if the pruner has been enabled (`ImagePrunerEnabled`), and perhaps another reporting if the most recent run was successful/failed (`ImagePruningSucceeded`)

gabemontero replied:
most recent run success/failed sounds good;

were you yay/nay on an "job in progress"? .... depends on how long prune jobs go for I suppose...

In theory, each config value under spec should have a corresponding value under status, especially if there are reasons a config setting would not be applied (like a job is in progress, and the job takes a long time).

adambkaplan commented:
@bparees @openshift/openshift-team-developer-experience updates to the enhancement proposal:

1. Moved the API to live in its own CRD. Reconciliation of this CRD is independent of the `ManagementState` set in the image registry operator's `ClusterOperator`.
2. Added a `status` to the API.

bparees commented:
it lgtm, will give @coreydaley and @dmage a chance to weigh in.

gabemontero commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,300 @@
+    lastTransitionTime: 2019-10-09T03:13:45
+    reason: Scheduled
+    message: "Image pruning has been scheduled."
+  - type: Completed
```
`InProgress` seems more intuitive to me, but I can defer if others do not agree

dmage replied:
Is `Completed` information about the latest successful run? I'd prefer not to have a condition that is by design going to periodically change its state.

adambkaplan replied:
> InProgress seems more intuitive to me, but I can defer if others do not agree

@gabemontero `Completed` comes from the Job API [1] - if we're bubbling up info from other API objects, I'd rather stay consistent.

> Is Completed information about the latest successful run? I'd prefer not to have a condition that is by design going to periodically change its state.

@dmage it is about the last run (success or failed). My intent was to make it easier to access this information, as `CronJob` does not directly indicate if the last run succeeded or not.

[1] https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#jobcondition-v1-batch

gabemontero replied:
ok works for me thanks

adambkaplan replied:
@dmage would it make more sense if we exposed the net status of the pruning jobs as Prometheus metrics? Ex - a gauge with the number of active pruning jobs (should be 1 or 0), and a counter of the completed pruning jobs by result (result="succeeded"|"failed"). The latter seems necessary if we want to alert on failure.

dmage replied:
@adambkaplan we definitely will want to have metrics and alerts. I can imagine two alerts:

1. the latest pruning job has finished unsuccessfully
2. the latest pruning job works for more than 1 day

I'm not sure about counters though, it might be hard to count jobs when they are getting pruned.

adambkaplan replied:
Prometheus counters are interesting - they work as an "increment only" indicator for things like HTTP requests [1]. On paper we could watch the pruning jobs in the namespace and update metrics.

In any case I don't think we should sweat the details on the metrics too much - during development we can decide which kind of metric is appropriate to drive a desired alert. I'll update the telemetry section with the alerts we want.

[1] https://prometheus.io/docs/concepts/metric_types/#counter

dmage replied:
Regarding conditions, I'm ok with having conditions about the latest success/failure, but the name `Completed` is ambiguous for our resource, it works only for Jobs (they can have either `Completed` or `Failed`). Upstream decided not to propagate information from Jobs to CronJobs, so we don't have an easy answer for this. I'd rather have a single condition whether the latest run is successful or not.

adambkaplan replied:
Restored the `Failed` condition so we can provide visibility on pruning job failures.

gabemontero commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,300 @@
+    lastTransitionTime: 2019-10-09T03:13:45
+    reason: "{reason from most recent pruning job}"
+    message: "{message from most recent pruning job}"
+  - type: Failed
```
Similarly, `Succeeded`

adambkaplan replied:
Per prior discussion - `Failed` aligns with Job API.

bparees commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,300 @@
+running.
+2. `suspend` - based on the provided `schedule`. If empty `suspend = true`.
+3. Pruning CLI:
+   1. `--prune-registry` - always `true` (default)
```
what effect does setting this to true have if the registry is not installed/present?  does it matter?

dmage replied:
Apparently it doesn't. We can defer pruning without the registry until someone need it.

bparees replied:
it doesn't what?

if someone has no registry, will this feature still be able to clean up their etcd (imported images)?

dmage replied:
No, `oc adm prune images` doesn't work without the registry and we need to fix it first: https://github.com/openshift/oc/blob/b6f4792b69cc6cb9d07f8ba88e73831c8a8fc470/pkg/cli/admin/prune/images/images.go#L430-L436
And that answers the question if someone uses it without the registry.

bparees replied:
yes, well that definitely answers that :)

can we get a BZ open to track fixing that?

adambkaplan replied:
Done - https://bugzilla.redhat.com/show_bug.cgi?id=1762837

adambkaplan commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,300 @@
+#### Spec
+
+1. `schedule:` Cron formatted schedule, defaults to daily at midnight for new
+clusters. If empty, job is suspended. Required field.
```
My experience with OpenAPI semantics is that `""` is an acceptable value for a required string field. Per the `CronJob` API `schedule` is a required field [1], meaning that clients must set a `schedule` value on creation.

In conversation with @coreydaley, he recommended that we align our API with the `CronJob` API if we are passing through fields. In that regard he proposed we add `suspend` as well as `startingDeadlineSeconds`. `suspend` is optional for the CronJob API, and keep it separate means that the pruner can be stopped without losing the prior schedule.

[1] https://github.com/kubernetes/api/blob/master/batch/v1beta1/types.go#L95

bparees replied:
i think @dmage's question was more like "we don't we make it optional, and then if it's provided, require the value to be a valid cron schedule(non-empty), and treat empty as disabled"?

now if CronJob also makes this field required, and allows it to be empty, then fine i guess...

bparees replied:
(if that wasn't @dmage's question, then it's my question :) )

adambkaplan commented:
@bparees @openshift/openshift-team-developer-experience think this is now in a final "implementable" draft. PTAL - thanks everyone for the detailed and thorough feedback!

bparees commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,325 @@
+
+1. `concurrencyPolicy: Forbid` - this prevents concurrent cron jobs from
+running.
+2. `suspend` - based on the provided `schedule`. If empty `suspend = true`.
```
this seems to conflict with the api described above which says the schedule must be set, and suspend is a valid field to configure.

bparees commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,325 @@
+  - Create a build with imagestream tag. Run the build.
+  - Delete build & IST.
+  - Run pruning.
+  - Verify image is pruned.
```
this is only going to prune the registry storage, right?  how will you verify the imagestream etcd content is being pruned?

adambkaplan replied:
@dmage if we set the pruner to be very aggressive (delete any image, keep only 1 tag instance) and we have two builds push to the same IST, running pruning should delete the result of the first build, correct?

adambkaplan replied:
(in both etcd and on the registry)

dmage replied:
@adambkaplan correct. If the first image is not referenced and we use `keep-younger-than: 0s`, it should delete one item from ImageStreamTag's history. And as the new images have unique SHA, the pruner should delete its Image resource too.

bparees commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,325 @@
+
+#### Tech Preview
+
+Requirements for reaching tech preview:
```
you're not targeting TP right? should this section be removed/updated to be NA?

adambkaplan replied:
correct - particularly since TP adds additional engineering costs

bparees commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,325 @@
+For GA the pruner will be enabled and run daily at midnight.
+
+On upgrade the cluster admin will be responsible for setting `suspend: false`
+on the pruning CRD. If pruning is suspended, an alert is fired.
```
```suggestion
on the pruning CRD in order to enable pruning. If pruning is suspended, an alert is fired.
```

adambkaplan commented:
@bparees @openshift/openshift-team-developer-experience updated with a more final form - ptal

bparees commented on image-registry/automate-pruning.md:
```diff
@@ -0,0 +1,336 @@
+
+The operator's behavior with respect to managing the pruner is independent of
+the `ManagementState` specified on the image registry operator's
+`ClusterOperator` object.
```
if the registry is "unmanaged" you'll still prune both etcd + storage?  I thought you were not going to prune storage in this case.

if the registry is "removed" you can't prune storage, so certainly that changes behavior?

adambkaplan replied:
We were only going to prune etcd if the registry is `Unmanaged` - I'll clarify this paragraph.

adambkaplan replied:
@bparees updated

bparees commented:
/lgtm
